### DISCUSSION TITLE
在服务端数据库的方案中，文档需要进一步说明避免用户落到坑里

### DISCUSSION THREAD

**Discussion by breakstring** [lobehub/lobehub#3062] [Self Hosting] [answered]
### 📦 部署环境

Vercel

### 📌 软件版本

1.1.6

### 💻 系统环境

macOS

### 🌐 浏览器

Edge

### 🐛 问题描述

在尝试现在的服务端数据库的方案时，遇到了好几个坑文档里面都没有提及。需要自己摸索（以及从 repo 里搜到相关的其他人的问题)才能解决。
1. 现有文档中貌似没有提到过，切换成服务端数据库后会导致原有客户端中的数据同步的数据全部丢失。因为切换成服务端数据库后，各个客户端也都同步切换了，需要登录，找不到原来的同步入口，导致我这几个月来的聊天记录完全失踪。。。
2. 希望文档中能够明确说明 S3 存储对于权限的需求。 文档对这块儿也写的比较粗略，我的部署过程中，一开始只是给了AK/AID 相关的全部读写部分，但是实际上还要开放对于这个桶的匿名读部分（这块儿对于客户端以及 OpenAI 等处理时也需要，一开始我以为咱们的客户端以及程序部分只是用 S3 来存储，对于客户端显示以及向服务端发送都会读取后编码 Base64 进行传输，看来不是这样。。。不开匿名的读，就无法显示也无法识别）。这块儿略有不安全，也请在文档中提及。
3. 现在的方案中似乎对于环境变量的依赖比较混乱，很多意想不到的纠缠。例如：
3.1 第一次尝试更改部署时，我只是尝试着加入了 Database_url 的环境变量，尚未配置 NEXT_PUBLIC_SERVICE_MODE=server 这个变量。本想着只是依次先把各种变量准备好后再切换成服务端。结果就导致重新部署后实际上已经变成服务端的机制了，看到后台有数据库 migrate 失败的提示。
3.2 等到我的 S3 等都准备好，进行第二次重新部署，发现都部署好后无法上传图片，上传图片时客户端/浏览器会报错。搜了搜 repo 的 issues 时，发现居然要删除掉原有的 ACCESS_CODE 这个环境变量。。。从字面意义上来看，这个环境变量和这部分功能并无直接关系。 删除环境变量，重新部署，随后一切正常。（不，还有刚才第二点说的权限问题，那个是摸索猜测才搞定）。

总之，感谢团队这次的服务端数据库的更新，但是在文档上还需要一些加强，否则对于像我们不熟悉代码结构的用户来说还是难免会出现各种问题。

### 🚦 期望结果

完善文档。

[a comment by an automated account is not shown]

**Comment by arvinxx** [comment]
1. 去掉服务端环境变量，还会回来的，可以试一下。另外文档开头已经提示了要存备份，你没有看到吗？
2. 这个权限和不同s3服务有关系吧，没法给详细文档，比如 R2 就没有这个配置，或者你可以 PR 一下你用的服务的说明？

3.1. 这一点你不完全按照文档指示，自己瞎搞为啥还要赖文档没写清楚… 如果没用，为啥我要把这个变量写上？😅
3.2. 这个应该是 bug ，后续会修一下的

本身服务端数据库就是一个比较复杂的部署方案。我问过好几个玩 self-hosting 的用户，他们甚至觉得现在的文档写的过于详细了。所以目前应该应该没什么需要再修改的。

如果有用户觉得现在这个文档还是过于简略，出了问题不知道该怎么解决，那我还是推荐用我们 hosting 的版本吧，省心一些🥲

[a comment by an automated account is not shown]

**Comment by breakstring** [comment]
> 1. 去掉服务端环境变量，还会回来的，可以试一下。另外文档开头已经提示了要存备份，你没有看到吗？
> 2. 这个权限和不同s3服务有关系吧，没法给详细文档，比如 R2 就没有这个配置，或者你可以 PR 一下你用的服务的说明？
> 
> 3.1. 这一点你不完全按照文档指示，自己瞎搞为啥还要赖文档没写清楚… 如果没用，为啥我要把这个变量写上？😅 3.2. 这个应该是 bug ，后续会修一下的
> 
> 本身服务端数据库就是一个比较复杂的部署方案。我问过好几个玩 self-hosting 的用户，他们甚至觉得现在的文档写的过于详细了。所以目前应该应该没什么需要再修改的。
> 
> 如果有用户觉得现在这个文档还是过于简略，出了问题不知道该怎么解决，那我还是推荐用我们 hosting 的版本吧，省心一些🥲

关于第一点。。。懒得再切换了，太折腾。。。
第二点，目前我使用的华为云的 OBS，和阿里云的 OSS 比较类似，他们对于 S3 类存储的权限控制的都很细致。不太清楚 AWS 和 CF 的 S3 存储的权限控制粒度会是怎样。至少在华为OBS 和阿里的 OSS 里面，如果默认的私有库，只开放了对应 AID/AK 的权限，图片能上传，但是 lobe-chat 的客户端以及通过 URL 方式发送到 OpenAI 之类的 AI 端点是无法访问到的，因为私有桶为了安全默认是无法通过 URL 来访问到里面的内容。正如上面所说的，对于我们不了解您这边详细代码的人来说很容易忽略这一点。我就以为你们是完全通过 base64 编码来向外发送以及在客户端展现。
关于 3.1。在文档里面的示例如下：
```
# Serverless Postgres DB Url
DATABASE_URL=

# 指定 service mode 为 server，否则不会进入服务端数据库
NEXT_PUBLIC_SERVICE_MODE=server
```
我根本就**没有设置 NEXT_PUBLIC_SERVICE_MODE=server 这个变量**，部署时实际上就进入了数据库的迁移逻辑。

[a comment by an automated account is not shown]

**Comment by arvinxx** [comment]
> 第二点，目前我使用的华为云的 OBS，和阿里云的 OSS 比较类似，他们对于 S3 类存储的权限控制的都很细致。不太清楚 AWS 和 CF 的 S3 存储的权限控制粒度会是怎样。至少在华为OBS 和阿里的 OSS 里面，如果默认的私有库，只开放了对应 AID/AK 的权限，图片能上传，但是 lobe-chat 的客户端以及通过 URL 方式发送到 OpenAI 之类的 AI 端点是无法访问到的，因为私有桶为了安全默认是无法通过 URL 来访问到里面的内容。正如上面所说的，对于我们不了解您这边详细代码的人来说很容易忽略这一点。我就以为你们是完全通过 base64 编码来向外发送以及在客户端展现。

可能这一部分是需要补充说明下。刚看了下文档里关于 S3 图片的使用在这里一笔带过了：

<img width="888" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/a0fc7fec-e3d1-42ec-9961-03640bad68ab">

在客户端 db 的情况下是都通过 base64 传输的，但是在服务端 db 模式中是直接传图片 url。其实后者才是 OpenAI 最推荐的方案（[文档](https://platform.openai.com/docs/guides/vision/managing-images)），可以大大降低请求传输的流量开销，响应速度也会更快。只是之前在没有服务端 db 的情况下，只能用 base64 编码。现在有 server db 了，自然走图片 URL 更好。

> 我根本就没有设置 NEXT_PUBLIC_SERVICE_MODE=server 这个变量，部署时实际上就进入了数据库的迁移逻辑。

数据库迁移和应用是两个不同的逻辑。数据库迁移就只依赖 DATABASE_URL 的，而且只在构建时执行，这个逻辑没问题。如果你没配置 NEXT_PUBLIC_SERVICE_MODE ，构建出来的应用就还是 client db 的。

[a comment by an automated account is not shown]

**Comment by ZeroDeng01** [comment]
> > 1. 去掉服务端环境变量，还会回来的，可以试一下。另外文档开头已经提示了要存备份，你没有看到吗？
> 
> > 2. 这个权限和不同s3服务有关系吧，没法给详细文档，比如 R2 就没有这个配置，或者你可以 PR 一下你用的服务的说明？
> 
> > 
> 
> > 3.1. 这一点你不完全按照文档指示，自己瞎搞为啥还要赖文档没写清楚… 如果没用，为啥我要把这个变量写上？😅 3.2. 这个应该是 bug ，后续会修一下的
> 
> > 
> 
> > 本身服务端数据库就是一个比较复杂的部署方案。我问过好几个玩 self-hosting 的用户，他们甚至觉得现在的文档写的过于详细了。所以目前应该应该没什么需要再修改的。
> 
> > 
> 
> > 如果有用户觉得现在这个文档还是过于简略，出了问题不知道该怎么解决，那我还是推荐用我们 hosting 的版本吧，省心一些🥲
> 
> 
> 
> 感觉目前文档确实有点 conflicted. 我昨晚刚从 0.x 操作升级到 1.x 的有服务端 persistence 的配置。
> 
> SSO 部分比较顺滑，得益于确实配置非常简洁；但 DB 部分。。至今也没搞定持久化。
> 
> 
> 
> 文档的一些问题：
> 
> 1. 环境变量几个地方不统一。ENABLE_OAUTH_SSO 在某页说必填，但实际上没有也行。NEXT_AUTH_SSO_PROVIDERS 在 `环境变量`部分的汇总说明提示可以有多个，但是之前的独立配置没提。https://lobehub.com/zh/docs/self-hosting/environment-variables  这个汇总说明里，没有提数据库部分。至少我对文档的阅读习惯，由于我 docker compose 部署的，我是上来就先看这页的，但一头雾水。另外印象中还有某页把 NEXT_AUTH_SECRET 写成了 NEXTAUTH_SECRET? 忘了，反正昨晚就是一团混乱，最后还是通过 docker logs 看启动日志里提到有一些变量 deprecated, 然后改过来
> 
> 2. 关于数据库部分，我觉得我似乎是配置了，但没有生效。。但 docker logs 也没说到底用没用 server side pg db. 我现在是不急用，否则真是要心急如焚呀
> 
> KEY_VAULTS_SECRET=
> 
> DATABASE_URL=
> 
> DATABASE_DRIVER=node
> 
> NEXT_PUBLIC_SERVICE_MODE=serve
> 
> 这几项都正确配置了，跨浏览器登录，对话还是没同步。
> 
> 
> 
> 至于 ACCESS_CODE 还要不要，现在也不明白。。

docker部署需要自己修改dockerfile然后build，官方目前的docker镜像由于next环境变量和构建问题不支持db版

[a comment by an automated account is not shown]

**Comment by arvinxx** [comment]
> 文档是不是没提

文档开头就说了不支持 Docker ，你是没看到么…

[a comment by an automated account is not shown]

**Comment by arvinxx** [comment]
> 作者既然说了那肯定是有，但我确实不知道哪一页说了，我的错。

我感觉你大概率就是没看这个： https://lobehub.com/zh/docs/self-hosting/advanced/server-database ，直接上手配。这样不出问题才怪 🤦‍♂️

[a comment by an automated account is not shown]

**Comment by arvinxx** [comment]
> 另外，如果能通过修订 Dockerfile 自行 build 解决，能否哪怕在现有的 Dockerfile 加十几行注释，给不熟悉代码结构以及 node 的 seflhosted 用户指条路？

现在你配置 next-auth 是没法正常完成用 db 实现鉴权的，要用也只能用 clerk。然后 @cy948 正在做 next-auth 和db 的集成，到时候我会专门给 serverdb 打一个 docker 镜像，届时应该直接配置就好了。

如果你熟悉 dockerfile 会做 build 的话，那么只需要记住 `NEXT_PUBLIC_` 前缀的环境变量，都需要放在 builder 阶段去进行配置。其他非 `NEXT_PUBLIC_` 前缀的变量都可以在运行时配置。

[a comment by an automated account is not shown]

**Comment by RubuJam** [comment]
但或许现在你还可以找回你原来的聊天记录，去回档到未启用clerk前的部署，然后导出所有文件，再在新的部署上导入即可

It may be possible to retrieve your original chat history, revert to the deployment prior to the introduction of Clerk, export the files, and then import them to the new deployment.

**Comment by cy948** [comment]
@utccnu 自己fork一个版本，关掉typecheck和lintcheck，就像这样： https://github.com/cy948/lobe-chat/blob/5095b9fa5b2939d394b7b71b2f003b3173ced1ec/next.config.mjs#L58-L63 。关掉之后亲测 2C8G 的机器也能完成构建。

[a comment by an automated account is not shown]

**Comment by cy948** [comment]
@utccnu 看看日志？返回 code137 的话，应该没有内存溢出了吧？

[a comment by an automated account is not shown]

**Comment by cy948** [comment]
@utccnu 我最近也构建不成功，直接  heap out of memory 了

[a comment by an automated account is not shown]

**Comment by cy948** [comment]
> 我最近也构建不成功，直接 heap out of memory 了

刚在自己电脑上build成功了，用时6分钟。🫠

[a comment by an automated account is not shown]

**Comment by cy948** [comment]
@utccnu 不如加入 discord，试试使用 NextAuth 的后端数据库版本。

[a comment by an automated account is not shown]

**Comment by Sun-drenched** [comment]
> @utccnu 我最近也构建不成功，直接 heap out of memory 了

构建的话也不一定非要用自己的机器吧，用容器镜像服务就行了。

[a comment by an automated account is not shown]

**Comment by scrpr** [comment]
使用Node Postgres的时候，还需要注意自签名证书的问题
部分Postgres服务商使用自签名的CA证书，比如[Aiven](https://aiven.io/docs/platform/concepts/tls-ssl-certificates)

此时如果使用服务方提供的URL的话，就会出现自签名证书问题：
```
❌ Database migrate failed: Error: self-signed certificate in certificate chain
    at /vercel/path0/node_modules/pg-pool/index.js:45:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at PgDialect.migrate (/vercel/path0/node_modules/src/pg-core/dialect.ts:72:3)
    at Object.migrate (/vercel/path0/node_modules/src/neon-serverless/migrator.ts:10:2)
    at runMigrations (/vercel/path0/scripts/migrateServerDB/index.ts:12:3) {
  code: 'SELF_SIGNED_CERT_IN_CHAIN'
```

原因在于，node-postgres在解析`Connection String`的时候行为与官方文档不一致。官方文档里`sslmode=require`的时候，如果本地存在CA证书就检查证书，没有就不检查；
![a6e295c6b14482e1728536b4f70d5475](https://github.com/lobehub/lobe-chat/assets/46473171/bf932e4d-af02-4b60-8d97-f6195537cc8f)
而node-postgres的实现中，多出了官方文档中没有的`no-verify`，选择此项时则不会验证证书；而`prefer`、`require`、`verify-ca`、`verify-full` 这4个选项行为完全一致，都是验证所有证书。
![ac3b70478e83d0d2a3842016ddde3c7e](https://github.com/lobehub/lobe-chat/assets/46473171/63934c8e-650c-476c-b729-0dc658500c17)

早在2年前便已经有人提出了修复pr使行为对齐官方文档，但至今未合并。https://github.com/brianc/node-postgres/pull/2709

**太长不看**：node-postgres实现跟官方不一致，将URL中的`?sslmode=require`替换为`?sslmode=no-verify`即可。

**Comment by scrpr** [comment]
> > 使用Node Postgres的时候，还需要注意自签名证书的问题 部分Postgres服务商使用自签名的CA证书，比如[Aiven](https://aiven.io/docs/platform/concepts/tls-ssl-certificates)
> > 此时如果使用服务方提供的URL的话，就会出现自签名证书问题：
> > ```
> > ❌ Database migrate failed: Error: self-signed certificate in certificate chain
> >     at /vercel/path0/node_modules/pg-pool/index.js:45:11
> >     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> >     at PgDialect.migrate (/vercel/path0/node_modules/src/pg-core/dialect.ts:72:3)
> >     at Object.migrate (/vercel/path0/node_modules/src/neon-serverless/migrator.ts:10:2)
> >     at runMigrations (/vercel/path0/scripts/migrateServerDB/index.ts:12:3) {
> >   code: 'SELF_SIGNED_CERT_IN_CHAIN'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 原因在于，node-postgres在解析`Connection String`的时候行为与官方文档不一致。官方文档里`sslmode=require`的时候，如果本地存在CA证书就检查证书，没有就不检查； ![a6e295c6b14482e1728536b4f70d5475](https://private-user-images.githubusercontent.com/46473171/343582339-bf932e4d-af02-4b60-8d97-f6195537cc8f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0NTY0MTUsIm5iZiI6MTcxOTQ1NjExNSwicGF0aCI6Ii80NjQ3MzE3MS8zNDM1ODIzMzktYmY5MzJlNGQtYWYwMi00YjYwLThkOTctZjYxOTU1MzdjYzhmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDAyNDE1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZDFiNGMxMzA2NzZmY2UyM2QxNDhhMjNlNzZhMDBkOWU0ODAzZDM3Mzk3MDFiYTllYTAyMzQxYzljZjNlZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.en5rydgjk75-lwGHZi11eAA-V9PhUo8pn284NZFDHGk) 而node-postgres的实现中，多出了官方文档中没有的`no-verify`，选择此项时则不会验证证书；而`prefer`、`require`、`verify-ca`、`verify-full` 这4个选项行为完全一致，都是验证所有证书。 ![ac3b70478e83d0d2a3842016ddde3c7e](https://private-user-images.githubusercontent.com/46473171/343582684-63934c8e-650c-476c-b729-0dc658500c17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0NTY0MTUsIm5iZiI6MTcxOTQ1NjExNSwicGF0aCI6Ii80NjQ3MzE3MS8zNDM1ODI2ODQtNjM5MzRjOGUtNjUwYy00NzZjLWI3MjktMGRjNjU4NTAwYzE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDAyNDE1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3ZDIyZTAyNDI1ODliNjYwM2IyMzA3ZTlhMTRmYTg4YTBlMDIyYmYwM2IzNzBjNTVkZTdiYmU1ZTExMjM1YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZXZDpYK65Cmb1p4UbDfewTbekUVv08BN_3JQ5N4Oiec)
> > 早在2年前便已经有人提出了修复pr使行为对齐官方文档，但至今未合并。[brianc/node-postgres#2709](https://github.com/brianc/node-postgres/pull/2709)
> > **太长不看**：node-postgres实现跟官方不一致，将URL中的`?sslmode=require`替换为`?sslmode=no-verify`即可。
> 
> 刚才我也试了 aiven, 同样是我上面提到的 443 访问错误。。你 aiven build 能通过？

非serverless的postgres实例需要添加环境变量更改driver，参见官方文档（现在访问不了服务端数据库的页面？）
```
  # 指定 Postgres database driver 为 node
  DATABASE_DRIVER=node
```

**Reply by Yuyuyang** [reply]
非常感谢这个讨论，我遇到了一样的问题！

**Comment by arvinxx** [comment, marked as answer]
https://lobehub.com/docs/self-hosting/server-database 目前初步调整完毕服务端数据库的部署文档框架，相关内容正在陆续补充中。有任何问题欢迎在 https://github.com/lobehub/lobe-chat/issues/3391 回复
